### ISSUE TITLE
Auction scan issue

### ISSUE THREAD

**Issue by ProfessorMinius** [andrew6180/TradeSkillMaster#28]
Hi, I'm getting this log after a scan finishes. It also stays forever on the last page of scanning:

Addon: TradeSkillMaster_AuctionDB
Message:  Interface\AddOns\TradeSkillMaster\Core\Threading.lua:28: ...SkillMaster_AuctionDB\TradeSkillMaster_AuctionDB.lua:398: table index is nil
Date: 01/05/23 04:35:41
Client: 3.3.5
Locale: enUS
Stack:
[C]: <pcall>
[string "*:OnLoad"]:4: <[string "*:OnLoad"]:1>
[C]: <error>
TSM\Core\Threading.lua:28: <callback>
TSM\Util\Delay.lua:30: <Delay.lua:18>
TSM Stack:
TSM_AuctionDB:DecodeItemData(4385)
Local Variables:

TSM Event Log:

Addons:
    AutoVendor (v0.9.10)
    Genie (3.0.5.1_beta2)
    TSM (v2.8.3)
    TSM_Accounting (v2.3)
    TSM_AuctionDB (v2.3.10)
    TSM_Auctioning (v2.3.3)
    TSM_Crafting (v2.5.2)
    TSM_Destroying (v2.1.1)
    TSM_ItemTracker (v2.0.8)
    TSM_Mailing (v2.2.1)
    TSM_Shopping (v2.3.6)
    TSM_Warehousing (v2.0.10)
    ElvUI (6.09)
    ElvUI_OptionsUI (1.06)

**Comment by ProfessorMinius** [comment]
Update: I removed the item with the id 4385 (Green Tinted Goggles) from my lua files and had to delete the DB lua, but there is no error now. Why would some random item be causing this?

**Comment by Bananaman** [comment]
Alright, so... thanks a lot for providing the call stack!

By looking at the code, we can see that it's attempting to decode the compact scan data from your database, and is failing because the `day` variable is becoming `nil` (empty), which would happen if there's corrupt data stored in the database:

https://github.com/andrew6180/TradeSkillMaster/blob/eb6607072a74b7156b5261a3a2b4c71aa025281d/TradeSkillMaster_AuctionDB/TradeSkillMaster_AuctionDB.lua#L390-L398

But why is there corrupt data in the database? Who knows. Probably impossible to find out, because this old TSM 2.8 code is a buggy mess, so it could have happened anywhere in the code!

Next, we can see that this `decodeScans` function is called by the following function, thanks to the call stack you provided:

https://github.com/andrew6180/TradeSkillMaster/blob/eb6607072a74b7156b5261a3a2b4c71aa025281d/TradeSkillMaster_AuctionDB/TradeSkillMaster_AuctionDB.lua#L468-L479

The rest of the call stack is missing, because this code is being executed via an anonymous callback, so it's impossible to say which function originally started the attempt to decode item data from the archive database.

So, all we can do is think a bit:

- Something is corrupting an entry in your database. It's impossible to say what did it.
- We can't see what the rest of the call stack is, so we can't think about how the data is being used. But I know from experience that `TSM:DecodeItemData` is the main function which converts TSM's "archive database" data to readable item data (their last-seen price, last scan date, etc).
- So... we know that we're unable to decode certain days sometimes, due to corrupt data in the SavedVariables.

With all these facts, my only conclusion is that we should add a simple protection in the data decoder, so that "if day is nil, skip this day". We can already see that they're skipping days in some other cases when other data is nil, so it makes sense to skip if the day itself is somehow corrupted.

I can also see some other good news: When `decodeScans()` runs, it replaces the cached "scans (scanned days)" data for that item in-memory, which means that the next time that something calls `TSM:EncodeItemData`, it will encode the cleaned-up "scans" array instead.

In other words... I know a safe fix for this bug. I do not know (and nobody will ever know) why your data became corrupted in the first place, but it doesn't really matter since it seems to be something extremely rare, and losing occasional days due to item corruption would be a small price to pay to fix the majority of this bug.

I'm going to add a fix for this issue in my pull request.

**Comment by Bananaman** [comment]
@ProfessorMinius By the way, did you ever create auction scan data with another version of TSM? That might be one reason for "corrupt" data, because it simply stored it in a different format. I don't think you've done that, but I have to ask. :)

**Comment by ProfessorMinius** [comment]
> Alright, so... thanks a lot for providing the call stack!
> 
> By looking at the code, we can see that it's attempting to decode the compact scan data from your database, and is failing because the `day` variable is becoming `nil` (empty), which would happen if there's corrupt data stored in the database:
> 
> https://github.com/andrew6180/TradeSkillMaster/blob/eb6607072a74b7156b5261a3a2b4c71aa025281d/TradeSkillMaster_AuctionDB/TradeSkillMaster_AuctionDB.lua#L390-L398
> 
> The error is that `day` is nil, which means that `scans[day]` is an illegal key (as it says in your log: `TradeSkillMaster_AuctionDB.lua:398: table index is nil`).
> 
> But why is there corrupt data in the database? Who knows. Probably impossible to find out, because this old TSM 2.8 code is a buggy mess, so it could have happened anywhere in the code!
> 
> Next, we can see that this `decodeScans` function is called by the following function, thanks to the call stack you provided:
> 
> https://github.com/andrew6180/TradeSkillMaster/blob/eb6607072a74b7156b5261a3a2b4c71aa025281d/TradeSkillMaster_AuctionDB/TradeSkillMaster_AuctionDB.lua#L468-L479
> 
> The rest of the call stack is missing, because this code is being executed via an anonymous callback, so it's impossible to say which function originally started the attempt to decode item data from the archive database.
> 
> So, all we can do is think a bit:
> 
> * Something is corrupting an entry in your database. It's impossible to say what did it.
> * We can't see what the rest of the call stack is, so we can't think about how the data is being used. But I know from experience that `TSM:DecodeItemData` is the main function which converts TSM's "archive database" data to readable item data (their last-seen price, last scan date, etc).
> * So... we know that we're unable to decode certain days sometimes, due to corrupt data in the SavedVariables.
> 
> With all these facts, my only conclusion is that we should add a simple protection in the data decoder, so that "if day is nil, skip this day". We can already see that they're skipping days in some other cases when other data is nil, so it makes sense to skip if the `day` variable itself is somehow corrupted.
> 
> I can also see some other good news: When `decodeScans()` runs, it replaces the cached "scans (scanned days)" data for that item in-memory, which means that the next time that something calls `TSM:EncodeItemData`, it will encode the cleaned-up "scans" array into your saved database instead. Therefore, as long as we clean up the "scans" data when we're loading it, we'll be able to save clean data to disk afterwards.
> 
> In other words... I know a safe fix for this bug. I do not know (and nobody will ever know) why your data became corrupted in the first place, but it doesn't really matter since it seems to be something extremely rare, and losing occasional days due to item corruption would be a small price to pay to fix the majority of this bug.
> 
> I'm going to add a fix for this issue in my pull request.

Thank you for your time and energy! At least now I won't lose all of my data because of a corrupted day.

> @ProfessorMinius By the way, did you ever create auction scan data with another version of TSM? That might be one reason for "corrupt" data, because it simply stored it in a different format. I don't think you've done that, but I have to ask. :)

I haven't done that, I think. I've only ever used this version of TSM on my client and it simply tends to happen on a whim.

**Comment by Bananaman** [comment]
@ProfessorMinius Hi again! I've now forked the project and created "TradeSkillMaster: Revived", which fixes an absolute ton of serious bugs and adds lots of new features. The download link and description is at the following URL:

https://github.com/Bananaman/TradeSkillMaster/releases

As the instructions say, I'd highly recommend deleting your old "AuctionDB" savedvariables, since the old TSM generates corrupt and incorrect market values. Your old database is useless, since it only contains market data, and it's all wrong since it was generated with an old TSM version.

Have fun and enjoy the game! :)

**Comment by Bananaman** [comment]
@ProfessorMinius If you somehow already fetched the release in the hour that's passed, you should fetch the zip again. I just noticed that TSM's "theme/color" switcher was resetting the custom fonts, because they had implemented it by erasing the entire design settings table (figures!). I've re-uploaded the release with a fix for that too. So now you can switch fonts and then switch color themes to your heart's content too! ;)

**Comment by ProfessorMinius** [comment]
@Bananaman wonderful news, I'll be sure to go move to your project!

**Comment by Bananaman** [comment]
@ProfessorMinius @DimaSheiko @akidraham @shravanshetty1 Apologies, I won't be pinging for any future updates after this, since that's unsustainable, but I figured some of you may want the latest release which came out today and has simply perfected the "time remaining" estimates for the "Full Scan" feature, and added a nice "total time elapsed" display when you finish a "Full Scan" or "Group Scan". I really don't anticipate anything else at the moment, and won't be able to keep alerting you about updates.

If you want news about future updates, go to the latest [TSM: Revived](https://github.com/Bananaman/TradeSkillMaster/releases/latest) release, then click on **"Watch"** at the top of the project page, then click on **"Custom",** and finally enable **"Releases",** then click on **"Apply"** to enable the notifications. Then GitHub will notify you if there are new releases in the future.

Take care everyone and have a great time in WoW! Good luck with your gold making! ;)

**Comment by DimaSheiko** [comment]
Thanks @Bananaman, I see you put a lot of effort into this, after you said you wouldn't, I still haven't seen the changes or tested them, I haven't had much free time lately, but eventually I will. Great job

**Comment by Bananaman** [comment]
@DimaSheiko Thanks, yeah I didn't intend to do such a massive rewrite, but as a fellow programmer you've probably seen this exact scenario before: Open a random project's codebase to fix 1 bug for yourself. Discover 1000000 bugs. It led to a long chain of fixes, along with some improvements I'd wanted.

The goal is just to have a reliable, "as perfect as possible" TSM for Wrath of the Lich King, and that goal is complete now. I definitely intend to maintain it "as necessary" for as long as I play. If there are any annoyances or bugs I notice and that affect the use of TSM for me, I'll fix them. But right now, everything I'm using TSM for works perfectly and better than ever before. ;)

Take care and have fun when you finally have time to check it out! :)
